whatweb

```diff
@@ -973,102 +935,93 @@ $CUSTOM_HEADERS.delete_if {|k,v| v=="" }
 
 ### TARGETS
 # clean up urls, add example urls if needed
-$targets=make_target_list(ARGV, input_file, $plugins_to_use)
-$recent_targets=[]
+$targets = make_target_list(ARGV, input_file, $plugins_to_use)
 
 # fail & show usage if no targets.
-if $targets.size <1
-	error "No targets selected"		
-	exit 1
+if $targets.size < 1
+  error("No targets selected")
+  exit 1
 end
 
-$semaphore=Mutex.new
-Thread.abort_on_exception = true if $WWDEBUG
-
-while t = next_target
-		Thread.new(t) do |thistarget|
-			begin
-				target = Target.new(thistarget) # we set the target within the thread
-			rescue => err
-				error(err)
-				next
-			end
-
-			puts Thread.current.to_s + " started for " + target.to_s if $verbose>1
-			sleep $WAIT unless $WAIT.nil? # wait
-
-			# follow redirects
-			no_redirects =false
-			num_redirects = 0
-			while no_redirects == false do
-				no_redirects=true if target.is_file?
-				# if we redirect 10 times we give up
-				if num_redirects == $MAX_REDIRECTS
-					error("ERROR Too many redirects: #{target.to_s}")
-					no_redirects=true
-					next
-				end
-
-				begin
-					target.open
-				rescue => err
-					error("ERROR Opening target: #{target.to_s} - #{err}")
-					no_redirects = true # without this we can get stuck in a loop
-					raise if $WWDEBUG
-					next
-				end
-
-				if target.is_url? and target.status.nil?
-					# assume all HTTP sites return a status
-					no_redirects=true
-					next
-				end
-
-				results = run_plugins(target)
-
-				# reporting
-				# multiple output plugins simultaneously, some stdout, some files
-				output_list.each do |o|
-					begin
-						o.out(target, target.status, results)
-					rescue => err
-						#srsly, logging failed
-						error("ERROR Logging failed: #{target.to_s} - #{err}")
-						raise if $WWDEBUG==true
-					end
-				end
-
-				# REDIRECTION
-				unless no_redirects
-					begin
-						if newtarget = target.get_redirection_target		
-							num_redirects+=1 
-							target=Target.new(newtarget)
-						else
-							no_redirects=true 
-						end
-					rescue => err
-						error("ERROR Redirection broken: #{target.to_s} - #{err}")
-						no_redirects=true
-						raise if $WWDEBUG==true
-					end
-				end
-			end # while no_redirects
-		end # Thread.new
-
-	while Thread.list.size>($MAX_THREADS+1)
-		puts "Thread list full, passing control" if $verbose>1
-		#sleep 0.5
-		Thread.pass
-	end
-end # targets.each
+# try to make a new Target object, may return nil
+def prepare_target(url)
+  begin
+    Target.new(url)
+  rescue => err
+    error(err)
+    nil
+  end
+end
 
-# close output logs
-output_list.each {|o| 
-	o.close
-}
+target_queue = Queue.new() # workers consume from this
```

Comment:
Using a queue allows for reading targets from a file (e.g. stdin) and processing the targets the targets at the same time.

Reply:
Cool!